src/librustc_resolve/late.rs

```diff
@@ -468,6 +468,18 @@ impl<'a, 'tcx> Visitor<'tcx> for LateResolutionVisitor<'a, '_> {
                 }
             }));
 
+        // rust-lang/rust#61631: The type `Self` is essentially
+        // another type parameter. For ADT's, we consider it
+        // well-defined only after all of the ADT type parameters have
+        // been provided. Therefore, we do not allow use of `Self`
+        // anywhere in type parameter defaults of ADT's.
+        //
+        // (We however cannot ban `Self` for defaults on *any* generic
+        // list; e.g. trait generics can usefully refer to `Self`)
```

Comment:
An example would be good here.

Reply:
Yes you're right; I put an example in the relevant unit test (issue-61631-default-type-param-can-reference-self-in-trat.rs), but it would be useful here as well.